types/applications/formApplication.d.ts

```diff
@@ -6,9 +6,13 @@
  * 2) The template used contains one (and only one) HTML form as it's outer-most element
  * 3) This abstract layer has no knowledge of what is being updated, so the implementation must define _updateObject
  *
+ * @typeParam D - The data structure used to render the handlebars template.
  * @typeParam O - the type of the object or entity target which we are using this form to modify
  */
-declare abstract class FormApplication<O = {}> extends Application {
+declare abstract class FormApplication<
+  D extends object = FormApplication.Data<{}>,
```

Comment:
Wouldn't it be easier to swap the generic parameters around, i.e.
```ts
declare abstract class FormApplication<
  O extends object = {},
  D extends object = FormApplication.Data<O>
> extends Application {
  /* ... */
}
```

Reply:
It would, but changing `D` is much more common than changing `O`.

Reply:
OK, I am coming from adifferent approach. I think we also need this for `BaseEntitySheet` and `ActorSheet` and `ItemSheet`. And in system development, the common use case is to change `O` as it is the entity.

Reply:
That makes sense. For those classes I think it would be reasonable to specify `O` first. I'll add a commit for that.